### PR TITLE
feat: update export feature to export the listed or filtered person

### LINKED ISSUE #153: feat: update export feature to only export filtered list

### PR DESCRIPTION
**Pull request by danielleloh**
for ease of use of the application, users should be able to directly export the filtered persons from the application instead of having to additionally filter the contacts in the csv file exported when all persons are exported.

## changes made
* add feature that only exports persons in the filteredPersonList of the addressbook.
* update user guide to match new behaviour of the export command (include steps to guide users to export filtered data)
* update test cases to test for new export feature that exports filtered persons list
* update developer guide to include changed made to the export command (sequence diagram and use case)

Screenshot 2024-03-23 at 10 31 18 PM

fixes #153

### CODE REVIEW

File: src/test/java/seedu/address/logic/commands/ExportCommandTest.java
Comment (Zer0Legion) [inline src/test/java/seedu/address/logic/commands/ExportCommandTest.java:1]: Insane change to a deceptively big task with such high speed. You are truly a coding weapon.

File: src/main/java/seedu/address/logic/commands/ExportCommand.java [hunk 1]
```diff
@@ -97,11 +154,27 @@ public JsonNode readJsonFile(File jsonFile) throws CommandException {
...
         return personsArray;
     }
 
+    /**
+     * Creates the directory for the CSV file if it does not exist.
+     *
+     * @param csvFile The CSV file for which the directory needs to be created.
+     * @throws CommandException If directory fails to be created.
+     */
+    public void createCsvDirectory(File csvFile) throws CommandException {
```
Comment (Zer0Legion) [inline src/main/java/seedu/address/logic/commands/ExportCommand.java:168]: Good handling of Boundary Value Analysis! Glad to see it being applied here after reading up on it this week.

File: docs/DeveloperGuide.md [hunk 2]
```diff
@@ -543,15 +543,25 @@ into user's clipboard.
...
 
-**Use case: UC10 — Export persons to CSV**
+**Use case: UC10 — Export listed persons to CSV**
 
 **MSS:**
 
-1.  User requests to export all persons and details to a CSV file.
-2.  AddressBook exports the persons to a CSV file.
-3.  AddressBook displays a message to confirm that all persons have been exported to a CSV file.
+1.  User !!requests to filter persons (UC06)!! by desired requirements
```
Comment (jayllo-c) [inline docs/DeveloperGuide.md:550]: Good job updating the DG to match the changes made

File: docs/UserGuide.md [hunk 3]
```diff
@@ -187,16 +187,22 @@ Format: `clear`
 
 ### Exporting Data to a CSV file : `export`
 
-Exports all persons and their details to a CSV file, addressbook.csv, which can be found in addressbookdata.
+Exports currently listed persons and their details to a CSV file, avengersassemble.csv, which can be found in addressbookdata.
```
Comment (jayllo-c) [inline docs/UserGuide.md:190]: Great job adding to the UG as well!

File: docs/diagrams/ExportSequenceDiagram.puml [hunk 4]
```diff
@@ -12,6 +12,11 @@ end box
```
Comment (jayllo-c) [inline docs/diagrams/ExportSequenceDiagram.puml:12]: Diagram looks correct!

File: src/main/java/seedu/address/logic/commands/ExportCommand.java [hunk 5]
```diff
@@ -140,15 +213,24 @@ public void writeToCsvFile(File csvFile, JsonNode jsonTree) throws IOException {
     public CommandResult execute(Model model) throws CommandException {
         requireNonNull(model);
 
-        Path jsonFilePath = model.getAddressBookFilePath();
-        File jsonFile = jsonFilePath.toFile();
-        File csvFile = new File(csvFilePath);
+        try {
+            ObservableList<Person> filteredPersonObservableList = model.getFilteredPersonList();
```
Comment (jayllo-c) [inline src/main/java/seedu/address/logic/commands/ExportCommand.java:217]: Nice job using the getFilteredPersonList() command to simplify the implementation of the code!

**Review by jayllo-c — COMMENTED** [review]
Great job altering the code to make it more useable for the target users!

File: src/test/java/seedu/address/logic/commands/ExportCommandTest.java [hunk 6]
```diff
@@ -3,102 +3,140 @@
...
-    public void execute_noPersonsArrayAddressBook_throwsCommandException() {
-        CommandException thrown = assertThrows(CommandException.class, () -> exportCommand
-                .execute(modelWithNoPersonsArray));
-        assertEquals("The JSON File is empty.", thrown.getMessage());
+    public void execute_validAddressBookExport_success() throws Exception {
+        CommandResult commandResult = exportCommand.execute(model);
+        assertEquals(exportCommand.MESSAGE_SUCCESS, commandResult.getFeedbackToUser());
+
+        File csvFile = new File("./src/test/data/ExportCommandTest/addressbookdata/addressbooktest.csv");
+        assertTrue(csvFile.exists());
```
Comment (delishad21) [inline src/test/java/seedu/address/logic/commands/ExportCommandTest.java:139]: Good extensive testing

File: src/main/java/seedu/address/logic/commands/ExportCommand.java [hunk 7]
```diff
@@ -140,15 +213,24 @@ public void writeToCsvFile(File csvFile, JsonNode jsonTree) throws IOException {
...
+            File jsonFile = filteredJsonFilePath.toFile();
+            File csvFile = new File(csvFilePath);
 
-        try {
             JsonNode jsonTree = readJsonFile(jsonFile);
             JsonNode personsArray = readPersonsArray(jsonTree);
+
+            createCsvDirectory(csvFile);
+
             writeToCsvFile(csvFile, personsArray);
```
Comment (delishad21) [inline src/main/java/seedu/address/logic/commands/ExportCommand.java:234]: Elegant way to incorporate getFilteredPersonList

**Review by delishad21 — COMMENTED** [review]
Elegant and neat solution to adding this feature. Well Done!

File: src/main/java/seedu/address/logic/commands/ExportCommand.java [hunk 8]
```diff
@@ -48,18 +63,60 @@ public void updateCsvFilePath(String filePath) {
...
+    }
+
+    /**
+     * Writes the information in the address book to a JSON file.
+     *
+     * @param jsonAddressBookStorage The JSON file to write information into.
+     * @param addressBook The address book from which to read the information from.
+     * @throws CommandException If information cannot be written into the JSON file.
+     */
+    public void writeToJsonFile(JsonAddressBookStorage jsonAddressBookStorage, AddressBook addressBook)
```
Comment (Pughal77) [inline src/main/java/seedu/address/logic/commands/ExportCommand.java:114]: Nice that you have used the classes already implemented in the storage package. This saves us a lot of time

File: src/test/java/seedu/address/logic/commands/ExportCommandTest.java (hunk 6 above)
Comment (Pughal77) [reply src/test/java/seedu/address/logic/commands/ExportCommandTest.java:139]: Yes this is very comprehensive

**Review by Pughal77 — COMMENTED** [review]
Nice clean implementation

**Pull request merged**